lib/ansible/modules/cloud/amazon/ec2_ami.py

```diff
@@ -542,69 +535,102 @@ def deregister_image(module, connection):
+        desired_groups = set(launch_permissions.get('group_names', []))
 
-    if image is None:
-        module.fail_json(msg="Image %s does not exist" % image_id, changed=False)
+        to_add_users = desired_users - current_users
+        to_remove_users = current_users - desired_users
+        to_add_groups = desired_groups - current_groups
+        to_remove_groups = current_groups - desired_groups
 
-    try:
-        set_permissions = connection.describe_image_attribute(Attribute='launchPermission', ImageId=image_id).get('LaunchPermissions')
-        if set_permissions != launch_permissions:
-            if ('user_ids' in launch_permissions or 'group_names' in launch_permissions):
-                group_names = launch_permissions.get('group_names')[0] if launch_permissions.get('group_names') else None
-                user_ids = launch_permissions.get('user_ids')[0] if launch_permissions.get('user_ids') else None
-                launch_perms_add = {'Add': [{}]}
-                if group_names:
-                    launch_perms_add['Add'][0]['Group'] = group_names
-                if user_ids:
-                    launch_perms_add['Add'][0]['UserId'] = user_ids
-                image.modify_attribute(Attribute='launchPermission', LaunchPermission=launch_perms_add)
-            elif set_permissions and set_permissions[0].get('UserId') is not None and set_permissions[0].get('Group') is not None:
-                image.modify_attribute(
-                    Attribute='launchPermission',
-                    LaunchPermission={
-                        'Remove': [{
-                            'Group': (set_permissions.get('Group') or ''),
-                            'UserId': (set_permissions.get('UserId') or '')
-                        }]
-                    })
-            else:
-                module.exit_json(msg="AMI not updated.", launch_permissions=set_permissions, changed=False,
-                                 **get_ami_info(get_image_by_id(module, connection, image_id)))
-            module.exit_json(msg="AMI launch permissions updated.", launch_permissions=launch_permissions, set_perms=set_permissions, changed=True,
-                             **get_ami_info(get_image_by_id(module, connection, image_id)))
-        else:
-            module.exit_json(msg="AMI not updated.", launch_permissions=set_permissions, changed=False,
-                             **get_ami_info(get_image_by_id(module, connection, image_id)))
-    except botocore.exceptions.ClientError as e:
-        module.fail_json(msg="Error updating image - " + str(e), exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+        to_add = [dict(Group=group) for group in to_add_groups] + [dict(UserId=user_id) for user_id in to_add_users]
+        to_remove = [dict(Group=group) for group in to_remove_groups] + [dict(UserId=user_id) for user_id in to_remove_users]
+
+        if to_add or to_remove:
```

Comment:
Not really sure what the desired behavior for this bit is. If I run a task with
```
        launch_permissions:
          group_names: ['all']
          user_ids: ['123456789012']
```
and a subsequent task removes the group_name 'all':
```
        launch_permissions:
          group_names: []
```
should that also remove the user_ids? That is what happens now.

Reply:
Yes, that would be what happens. If you pass launch permissions, you have to pass all the launch permissions. I honestly don't think many people modify launch permissions other than adding an AWS account. Could perhaps be better documented but that could be addressed later.

Reply:
I've updated the docs for launch permissions. I don't think the behaviour changes with this PR.